### DISCUSSION TITLE
HW1 Part 3

### DISCUSSION THREAD

**Discussion by tad12002** [cooperrc/computational-mechanics#192] [Q&A]
`dt = 20  
k=0.013
t_euler = np.arange(1900, 2021, dt)
p_euler = np.zeros(len(t))
p_euler[0] = pop[0]

for i in range(1, len(t)):
    p_euler[i] = p_euler[i-1] + k * p_euler[i-1] * dt`

error is in the for loop saying 'numpy.ndarray' object is not callable and I'm having issues in other question on part 3 any suggestions?

**Comment by cooperrc** [comment]
the function `range` is probably saved as a numpy array, e.g. 

`range = np.arange(20)`

**Reply by tad12002** [reply]
SO long story short, I did in fact assign a variable called range earlier in the code. When I ran the code separately from the hw in a new kernel it worked fine... So I sifted through the hw and found the issues yet it still had the error. So I re did the homework assignment putting all my work in a separate file to see if it would work and of course it did. I think what happened was the hw kernel stored a variable?

In short above leads to this question, do kernels store variables? Because after restarting the kernel everything worked fine but only did so until I restarted the kernel? Maybe I didn't run everything through properly? Finally is there a clear stored variables command like in matlab?

**Reply by cooperrc** [reply]
> do kernels store variables
Yes. Each Python session will have functions, libraries, and objects stored into variables for you to call on interactively. 

This is a double-edged sword:
- (positive) You can reuse variables+functions anywhere within a notebook (as long as you ran the cell already)
- (negative) If you reuse variable names, it will overwrite or create unexpected results when you run older cells e.g. 
```python
[In 1]: t = np.linspace(0, 1, 11)
        f = t**2
        plt.plot(t, f)
[In 2]: t = np.linspace(0, 10, 50)
        f2 = t**2
[In 3]: plt.plot(t, f)
...
ValueError: x and y must have same first dimension, but have shapes (50,) and (11,)

```
